test: Improve link monitor `ip_monitor_assert_stable_link_up`

cathay4t opened the pull request:
Current thread data sharing and function decorator is hard to understand,
Changed the ip monitor output to `/tmp/nmstate_test_ip_monitor.log` file
and check it after test finished.

For ethernet or veth, `nmcli c down` will not cause link down, hence
`ip_monitor_assert_stable_link_up()` changed to fail if specified
interface is ethernet or veth.

Changed test case `test_change_mtu_with_stable_link_up` to use linux
bridge instead of veth, so it will fails when nmstate is using normal
connection activation instead of reapply.

liangwen12year commented on tests/integration/testlib/iproutelib.py:
```diff
@@ -4,92 +4,66 @@
+            iface_state = show_only((dev,))[Interface.KEY][0]
+            assert iface_state[Interface.TYPE] not in [
+                InterfaceType.ETHERNET,
+                InterfaceType.VETH,
```
nit, you can ignore it though. 
```
            assert iface_state[Interface.TYPE] not in {
                InterfaceType.ETHERNET,
                InterfaceType.VETH,
            }, f"Device type {iface_state[Interface.TYPE]} is not allowed."
```

liangwen12year commented:
Approved it too soon, the integration test `test_modify_ipv6_with_reapply` is broken by the patch.

ihuguet commented on tests/integration/testlib/iproutelib.py:
```diff
@@ -4,92 +4,66 @@
-        self.out = None
-        self.err = None
-        self.popen = None
+IP_MONITOR_LOG_TMP_FILE = "/tmp/nmstate_test_ip_monitor.log"
```
Can tests be executed in parallel? Anyway, I don't see a good idea to use a globally defined file path. Either use pipes as before or use a file created like `mktmp`, IMHO.
Apart from that LGTM.

cathay4t replied:
Nmstate has lock on checkpoint, so tests should not be executed in parallel. Will change to mktemp.

cathay4t replied:
Changed to use `TemporaryFile()`.

tyll commented:
fixes #2811

cathay4t commented:
Test fixed. Please review again. Thanks.

cathay4t committed:
test: Improve link monitor `ip_monitor_assert_stable_link_up`

Current thread data sharing and function decorator is hard to understand,
Changed the ip monitor output to a temp file and check it after test finished.

For ethernet or veth, `nmcli c down` will not cause link down, hence
`ip_monitor_assert_stable_link_up()` changed to fail if specified
interface is ethernet or veth.

Changed test case `test_change_mtu_with_stable_link_up` to use linux
bridge instead of veth, so it will fails when nmstate is using normal
connection activation instead of reapply.

Changed test case in `static_ip_address_test.py` to use dummy interface
for link stability tests.

Signed-off-by: Gris Ge <fge@redhat.com>
